Upgrade vendored dependencies

sbidoul opened the pull request:
Upgrade vendored dependencies early in the 24.1 release cycle.

Missing here is certify (the patch does not apply), and of course, packaging.

closes #12556

pfmoore commented:
I don't think we should include the files being flagged in `msgpack` - they look like they are related to a compiled extension. I'm assuming `msgpack` can still be used as pure Python?

sbidoul commented:
I have opened https://github.com/msgpack/msgpack-python/issues/580 to enquire about this .cpp, .h and .pyx files that are now shipped in their wheels.

`msgpack` otherwise still works in pure python mode.

In the meantime I remove this commit.

sbidoul commented:
The last `typing-extension` that supports python 3.7 is 4.7.1, removing that commit.

@pradyunsg does the vendoring process have a way to check python version compatibility?

sbidoul commented:
Ditto for platformdirs. 

@pradyunsg another question, is it possible to specify an upper bound version for a package in vendoring process?

So, now python 3.7 support starts to be annoying. I'm now inclined to drop it.

pfmoore commented:
+1 on dropping 3.7 support.

pradyunsg commented:
> @pradyunsg another question, is it possible to specify an upper bound version for a package in vendoring process?

Oooh, not at the moment no.

pradyunsg commented:
> @pradyunsg does the vendoring process have a way to check python version compatibility?

No, it does not.

sbidoul commented:
I also updated `msgpack` (which has a release without the .cpp and .h files), `certifi` (I updated the patch) and `urliib3<2`.

Regarding urllib3, is there a particular reason to not upgrade to 2.x, beyond the fact that updating our patch is difficult?

pradyunsg commented:
> Regarding urllib3, is there a particular reason to not upgrade to 2.x,

IIRC, requests wasn't ready for the newer urllib3 at the time. We should probably look into upgrading to urllib3 unconditionally.

pradyunsg commented:
I merged this so that we don't end up with a long tail of conflicts on here -- let's tackle urllib3 in a separate PR. :)

notatallshaw commented:
The urllib3 2.0 issue was related to openssl support: https://github.com/pypa/pip/pull/12026#issuecomment-1630727697
